### ISSUE TITLE
Nullable DateTime in projection class - System.Data.SqlClient.SqlException: 'Conversion failed when converting date and/or time from character string.'

### ISSUE THREAD

**Issue by brunolau** [aspnet/EntityFrameworkCore#12797]
Having a one to many relationship, the main entity can have zero to n child entities. When I want to select the main entity's Id and most recent DateTime from the child entity collection, SQL execution crashes with a System.Data.SqlClient.SqlException

```
Exception message: 
Conversion failed when converting date and/or time from character string

Stack trace: 
   at System.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at System.Data.SqlClient.SqlInternalConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
   at System.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
   at System.Data.SqlClient.SqlDataReader.TryHasMoreRows(Boolean& moreRows)
   at System.Data.SqlClient.SqlDataReader.TryReadInternal(Boolean setTimeout, Boolean& more)
   at System.Data.SqlClient.SqlDataReader.Read()
   at Microsoft.EntityFrameworkCore.Storage.RelationalDataReader.Read()
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.Enumerator.BufferlessMoveNext(DbContext _, Boolean buffer)
   at Microsoft.EntityFrameworkCore.SqlServer.Storage.Internal.SqlServerExecutionStrategy.Execute[TState,TResult](TState state, Func`3 operation, Func`3 verifySucceeded)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.Enumerator.MoveNext()
   at System.Linq.Enumerable.FirstOrDefault[TSource](IEnumerable`1 source)
   at lambda_method(Closure )
   at Microsoft.EntityFrameworkCore.Query.Internal.LinqOperatorProvider.ResultEnumerable`1.GetEnumerator()
   at Microsoft.EntityFrameworkCore.Query.Internal.LinqOperatorProvider.ExceptionInterceptor`1.EnumeratorExceptionInterceptor.MoveNext()
   at System.Linq.Enumerable.First[TSource](IEnumerable`1 source)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass15_1`1.<CompileQueryCore>b__0(QueryContext qc)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.Execute[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.Execute[TResult](Expression expression)
   at System.Linq.Queryable.FirstOrDefault[TSource](IQueryable`1 source)
   at EfCoreErrReproduction.Program.Main() 


SQL produced:
Executed DbCommand (57ms) [Parameters=[@__id_0='3422'], CommandType='Text', CommandTimeout='30']
SELECT TOP(1) [p].[Id], (
    SELECT COUNT(*)
    FROM [TicketNumberDelta] AS [t]
    WHERE [p].[Id] = [t].[TicketNumberId]
) AS [DeltaCount], COALESCE((
    SELECT TOP(1) [lx].[RecordTimeStamp]
    FROM [TicketNumberDelta] AS [lx]
    WHERE [p].[Id] = [lx].[TicketNumberId]
    ORDER BY [lx].[RecordTimeStamp] DESC
), '0001-01-01T00:00:00.0000000') AS [LastEntry]
FROM [TicketNumber] AS [p]
WHERE [p].[Id] = @__id_0

System.Data.SqlClient.SqlException: 'Conversion failed when converting date and/or time from character string.'

```

### Steps to reproduce
[EfCoreErrReproduction.zip](https://github.com/aspnet/EntityFrameworkCore/files/2229460/EfCoreErrReproduction.zip)

```c#
using Microsoft.EntityFrameworkCore;
using System;
using System.Collections.Generic;
using System.Linq;

namespace EfCoreErrReproduction
{
    class Program
    {
        static void Main(string[] args)
        {
            var db = new StrippedDownContext();
            var crash = db.TicketNumbers.Where(p => p.Id == 3422).Select(p => new ProjectionClass
            {
                Id = p.Id,
                DeltaCount = p.TicketNumberDeltas.Count(),
                LastEntry = p.TicketNumberDeltas.OrderByDescending(lx => lx.RecordTimeStamp).Select(le => le.RecordTimeStamp).FirstOrDefault()
            }).FirstOrDefault();
        }
    }

    class ProjectionClass
    {
        public int Id { get; set; }
        public int DeltaCount { get; set; }
        public DateTime? LastEntry { get; set; }
    }
}

namespace EfCoreErrReproduction
{
    public class StrippedDownContext : DbContext
    {
        public StrippedDownContext()
        {
        }

        public StrippedDownContext(DbContextOptions<StrippedDownContext> options)
            : base(options)
        {
        }


        public virtual DbSet<TicketNumber> TicketNumbers { get; set; }


        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            if (!optionsBuilder.IsConfigured)
            {
                optionsBuilder.UseSqlServer("db conn string");
            }
        }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            modelBuilder.Entity<TicketNumber>(entity =>
            {
                entity.ToTable("TicketNumber");
            });

            modelBuilder.Entity<TicketNumberDelta>(entity =>
            {
                entity.ToTable("TicketNumberDelta");
               
                entity.HasOne(p => p.TicketNumber)
                    .WithMany(p => p.TicketNumberDeltas)
                    .HasForeignKey(p => p.TicketNumberId);

                entity.Property(e => e.RecordTimeStamp).HasColumnType("datetime");
            });
        }
    }

    public class TicketNumber
    {
        public TicketNumber()
        {
            TicketNumberDeltas = new List<TicketNumberDelta>();           
        }

        public int Id { get; set; }
        public virtual ICollection<TicketNumberDelta> TicketNumberDeltas { get; set; }
    }

    public partial class TicketNumberDelta
    {
        public int Id { get; set; }
        public int? TicketNumberId { get; set; }
        public DateTime RecordTimeStamp { get; set; }
        public virtual TicketNumber TicketNumber { get; set; }
    }
}
```

### Further technical details
EF Core version: 2.1.1
Database Provider: Microsoft.EntityFrameworkCore.SqlServer
Operating system: Win 10 Enterprise x64
IDE: Visual Studio 2017 15.7.5

**Comment by maumar** [comment]
I was able to reproduce this on current bits

**Comment by maumar** [comment]
@brunolau workaround is to cast the `RecordTimeStamp` to nullable `DateTime` before its projected, like so:

```cs
            var query = db.TicketNumbers.Where(p => p.Id == 3422).Select(p => new ProjectionClass
            {
                Id = p.Id,
                DeltaCount = p.TicketNumberDeltas.Count(),
                LastEntry = p.TicketNumberDeltas.OrderByDescending(lx => lx.RecordTimeStamp).Select(le => (DateTime?)le.RecordTimeStamp).FirstOrDefault()
            }).FirstOrDefault();
```

**Comment by brunolau** [comment]
Workaround works, thank you!

**Comment by ajcvickers** [comment]
@maumar Let's investigate for patch.

**Comment by ajcvickers** [comment]
See also #12819 when working on this.

**Comment by ajcvickers** [comment]
@maumar Any update on this?

**Comment by ksmithRenweb** [comment]
A potentially simpler reproduction is this.

var ret = from t in context.NullableDate
               select new NotNullableDateDto
               {
                   NotNullable = t.NullableDate ?? new DateTime(1900, 1, 1)
               }

Code worked fine in Ef Core 2.0

**Comment by maumar** [comment]
in 2.0 we used to generate:

```sql
SELECT COALESCE([t].[NullableDate], '1900-01-01T00:00:00.000') AS [NotNullable]
FROM [Nullables] AS [t]
```

and now we generate:
```sql
SELECT COALESCE([t].[NullableDate], '1900-01-01T00:00:00.0000000') AS [NotNullable]
FROM [Nullables] AS [t]
```

**Comment by smitpatel** [comment]
@ksmithRenweb - What is the data type of Column `NullableDate` in database?

**Comment by maumar** [comment]
@smitpatel  
```
            modelBuilder.Entity<Nullable>().Property(e => e.NullableDate).HasColumnType("datetime");
```

**Comment by maumar** [comment]
we should be using the type from the other side of coalesce to determine how to render a datetime literal for the second term. We have a general issue to be better about it, but perhaps there is something we can easily do for this specific case. @smitpatel thoughts?

**Comment by smitpatel** [comment]
https://github.com/aspnet/EntityFrameworkCore/blob/939f6d82e17c71f885419ff15e77b7ad8a0cda06/src/EFCore.Relational/Query/Sql/DefaultQuerySqlGenerator.cs#L1317-L1338

Looks like, our type mapping inference for Binary is missing out Coalesce by accident. Should be easy fix.

**Comment by ksmithRenweb** [comment]
@smitpatel Data Type of "datetime" and of course set to Allow Nulls.

**Comment by maumar** [comment]
@smitpatel  a disabled test has been added in 4cd4301c27ace85d24ef95dbb7eeee28d27b58d2 - when this is fixed make sure to enable the test

**Comment by ajcvickers** [comment]
Will discuss with @divega

**Comment by ajcvickers** [comment]
We should consider the general case here as part of #12795; specifically #12672

**Comment by smitpatel** [comment]
@ajcvickers to give this a read before next triage

**Comment by ajcvickers** [comment]
Notes from triage: The literal generator for the provider/type/value combination should throw if asked to generate a literal for values that cannot be represented in the store. Se also #12998.

**Comment by ajcvickers** [comment]
Removing from milestone to discuss in triage. Not sure what is left to do here.

**Comment by ajcvickers** [comment]
@smitpatel to create a repro.

**Comment by smitpatel** [comment]
In 3.0 query pipeline we reverted the fix we made for #10294 which introduced Coalesce function in sql tree. Now we let server side be null and when reading back valuetype from it, get default value. Hence we don't generate problematic SQL anymore.
